libbeat/registry/backend/memlog/util.go

```diff
@@ -0,0 +1,105 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package memlog
+
+import (
+	"hash"
+	"hash/fnv"
+	"io"
+	"os"
+	"reflect"
+	"syscall"
+	"unsafe"
+)
+
+var errno0 = syscall.Errno(0)
+
+type hashFn func(k []byte) uint64
+
+type ensureWriter struct {
+	w io.Writer
+}
+
+func (e *ensureWriter) Write(p []byte) (int, error) {
+	var N int
+	for len(p) > 0 {
+		n, err := e.w.Write(p)
+		N, p = N+n, p[n:]
+		if isRetryErr(err) {
```

Comment:
what if it's not retryable error? are we proceeding with writes as nothing happened?

Reply:
retrieable errors are EINTR, or EAGAIN only. `ensureWriter` is used to wrap a file writer, freeing us from checking and handling these two error conditions. All other IO errors should bubble up.